app/Services/ProvisioningService.php

```diff
@@ -0,0 +1,360 @@
+<?php
+
+namespace App\Services;
+
+use App\Enums\ServerStatus;
+use App\Enums\TimePeriod;
+use App\Models\Permission;
+use App\Models\Role;
+use App\Models\RoomType;
+use App\Models\Server;
+use App\Models\ServerPool;
+use App\Models\User;
+use App\Settings\GeneralSettings;
+use App\Settings\RecordingSettings;
+use App\Settings\RoomSettings;
+use App\Settings\UserSettings;
+use Illuminate\Database\RecordsNotFoundException;
+use Illuminate\Support\Facades\Validator;
+use Illuminate\Validation\Rule;
+use Log;
+use ReflectionClass;
+use UnexpectedValueException;
+use ValueError;
+
+abstract class AbstractProvisioner
+{
+    protected string $model;
+
+    protected array $expectedProperties;
+
+    protected function modelName()
+    {
+        $name = (new ReflectionClass($this->model))->getShortname();
+        $name = preg_replace_callback('/[A-Z]/', fn ($match) => ' '.strtolower($match[0]), $name);
+
+        return ltrim($name);
+    }
+
+    protected function createWrapper(object $properties, callable $callback)
+    {
+        $name = $properties->name ?? "$properties->firstname $properties->lastname";
+        Log::notice("Provisioning {$this->modelName()} '$name'");
+        $validator = Validator::make((array) $properties, $this->expectedProperties);
+        if ($validator->fails()) {
+            throw new UnexpectedValueException("Invalid {$this->modelName()} definition: {$validator->errors()}");
+        }
+        $item = new $this->model;
+        $callback($item);
+        $item->save();
+    }
+
+    protected function destroyWrapper(array $match, ?callable $callback = null)
+    {
+        if ($match) {
+            $expression = implode(' && ', array_map(fn ($a, $b) => "$a = $b", array_keys($match), array_values($match)));
+            Log::notice("Deleting all {$this->modelName()}s matching '$expression'");
+        } else {
+            Log::notice("Deleting all {$this->modelName()}s");
+        }
+        $items = $this->model::lazy();
+        foreach ($match as $key => $value) {
+            $items = $items->where($key, $value);
+        }
+        foreach ($items as $item) {
+            if ($callback) {
+                $callback($item);
+            }
+            if (! $item->delete()) {
+                $name = $item->name ?? "$item->firstname $item->lastname";
```

Comment:
Some models already have a method `getLogLabel` https://github.com/THM-Health/PILOS/blob/develop/app/Models/User.php#L80
You could use this and add it to all models you need and that still don't have it

Reply:
I like that. This would also solve the `$name` "problem" in `createWrapper`. Will do!

Reply:
Done.